src/operand/reference.ts

```diff
@@ -16,6 +16,16 @@ import { Operand } from '.'
  * @return {Result}
  */
 function contextValueLookup (ctx: Context, key: string): Result {
+  // Resolve complex keys
+  const complexKeyExpression = /{([^{}]+)}/
+  let complexKeyMatches = complexKeyExpression.exec(key)
+
+  while (complexKeyMatches) {
+    const resolvedValue = contextValueLookup(ctx, complexKeyMatches[1])
+    key = key.replace(complexKeyExpression, `${resolvedValue}`)
+    complexKeyMatches = complexKeyExpression.exec(key)
+  }
```

Comment:
thought: to make this lookup more effective, you really want to resolve the {} only for the given path segment (key[i]).
- e.g.: $a.{b}.{c}
- if the a would be resolved as undefined, you're returning right-away without the necessity to resolve the deeper references.

Reply:
pretty much the same thing you're doing for arrays

Reply:
another small tip/though :)
the nested replacement could be simplified to:

```
key = key.replace(/{([^{}]+)}/, (match, innerReferenceKey) => contextValueLookup(ctx, innerReferenceKey))
```

Reply:
Would it be possible for there to be more than 1 'inner reference' per path segment?

Reply:
It should be fine with "**g**" flag: https://regexr.com/59oa5 ... to sequentially resolve siblings pairs

Reply:
It fails on `Ref{Ref{RefE}}` though. And it doesn't handle the case where contextValueLookup returns undefined very well.

Reply:
Why it fails on ```Ref{Ref{RefE}}```?

undefined:

```
try {
   key = key.replace(/{([^{}]+)}/, (match, innerReferenceKey) => {
      const resolvedKey = contextValueLookup(ctx, innerReferenceKey)
      if (!resolvedKey) { throw new Error(`failed to resolve inner reference: "${innerReferenceKey}"`) }
   })
} catch { return undefined }

... take my code example really just as inspiration

Reply:
`Ref{Ref{RefE}}`

```
const ctx = {
  RefA: 'B',
  RefB: 'result',
  RefE: 'A',
  
}
```

If you only do one replace, it will only catch `{RefE}`. So the output will be `Ref{RefA}`.

If using a while loop, it will catch `{RefE}`, transform the key to `Ref{RefA}`, then on the next iteration it catches `{RefA}` and transforms the key to `RefB`

Also, even with the while loop. If I put it per path segment I'm having trouble with `RefG[{RefC.sub{RefD}}]`

Reply:
> `Ref{Ref{RefE}}`
> 
> ```
> const ctx = {
>   RefA: 'B',
>   RefB: 'result',
>   RefE: 'A'
> }
> ```
> 
> If you only do one replace, it will only catch `{RefE}`. So the output will be `Ref{RefA}`.
> 
> If using a while loop, it will catch `{RefE}`, transform the key to `Ref{RefA}`, then on the next iteration it catches `{RefA}` and transforms the key to `RefB`
> 
> Also, even with the while loop. If I put it per path segment I'm having trouble with `RefG[{RefC.sub{RefD}}]`
> 
> Edit:
> Also not a big fan of exceptions for flow control, unless it's in python :)

Make sense!

btw, here when **resolvedValue**  is undefined we should right-way return "undefined" no?
```
const resolvedValue = contextValueLookup(ctx, complexKeyMatches[1])
    key = key.replace(complexKeyExpression, `${resolvedValue}`)
```

Reply:
> ```
> const resolvedValue = contextValueLookup(ctx, complexKeyMatches[1])
>     key = key.replace(complexKeyExpression, `${resolvedValue}`)
> ```

Yes, that make sense. I realised my original code was doing that `Refundefined` as a string thing too, hehe.

Reply:
468e5f97a097c75b2e6d4465929dd0934c591405